### PR TITLE
Editorial: Clarify the AsyncGenerator state machine

### PR DESCRIPTION
**Pull request by bakkot**
AsyncGenerators are complicated, but at the root there's a relatively straightforward state machine. I find that it's difficult to see the shape of that machine with the current spec text, which routes every operation through [AsyncGeneratorResumeNext](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-asyncgeneratorresumenext).

I've rewritten all the logic here. The main reason I prefer this text is that each AO does one thing and returns, instead of kicking off a chain of self-recursive AOs which you have to follow for quite a while, keeping track of the state, before you figure out what's actually going to happen. For example, I think this PR makes it much more obvious that AsyncGeneratorYield might not return control to the rest of the program, and under what circumstances that happens. As a bonus, it makes AsyncGeneratorYield follow the pattern described in [this comment](https://github.com/tc39/ecma262/issues/2400#issuecomment-832335872), hopefully providing a path towards [clarifying execution context juggling](https://github.com/tc39/ecma262/issues/2409).

I don't believe there's any normative changes ~except for possibly, in certain cases, to the realm used for creating iterator result objects~. (Edit: I've tweaked things to ensure it preserves the original, fairly strange behavior. I think we may want to change that behavior, but not as part of this PR.)

This PR has an incorrect `!` assertion in callers of AsyncGeneratorAwaitReturn, corresponding to the same assertion in callers of AsyncGeneratorResumeNext described in https://github.com/tc39/ecma262/issues/2412. I'll fix this PR up if that issue is resolved first. Otherwise it just leaves things in the same state.

Rather than trying to confirm that the logic is the same by comparing the two texts, I strongly suggest you read [this gist](https://gist.github.com/bakkot/82186e27a2e78c6e507c56dd49d8a197) where I've described the state machine, and confirm that both the original spec text and the text in this PR correspond to that state machine.

cc @jmdyck, I know you've spent a long time puzzling over this as well.

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -39974,26 +39974,66 @@ <h1>AsyncGenerator.prototype.constructor</h1>
...
         <emu-alg>
           1. Let _generator_ be the *this* value.
+          1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+          1. Let _result_ be AsyncGeneratorValidate(_generator_, ~empty~).
+          1. IfAbruptRejectPromise(_result_, _promiseCapability_).
+          1. Let _state_ be _generator_.[[AsyncGeneratorState]].
+          1. If _state_ is ~completed~, then
+            1. Let _iteratorResult_ be ! CreateIterResultObject(*undefined*, *true*).
+            1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _iteratorResult_ &raquo;).
+            1. Return _promiseCapability_.
```
Comment (jmdyck) [inline spec.html:39984]: A function can't return a spec Record.

File: spec.html [hunk 2]
```diff
@@ -40105,76 +40149,43 @@ <h1>AsyncGeneratorValidate ( _generator_, _generatorBrand_ )</h1>
         </emu-alg>
       </emu-clause>
 
-      <emu-clause id="sec-asyncgeneratorresolve" aoid="AsyncGeneratorResolve">
-        <h1>AsyncGeneratorResolve ( _generator_, _value_, _done_ )</h1>
-        <p>The abstract operation AsyncGeneratorResolve takes arguments _generator_, _value_, and _done_ (a Boolean). It performs the following steps when called:</p>
+      <emu-clause id="sec-asyncgeneratorenqueue" aoid="AsyncGeneratorEnqueue">
+        <h1>AsyncGeneratorEnqueue ( _generator_, _completion_, _promiseCapability_ )</h1>
+        <p>The abstract operation AsyncGeneratorEnqueue takes arguments _generator_ (an Async Generator), _completion_ (a Completion Record), and _promiseCapability_ (a PromiseCapability Record). It performs the following steps when called:</p>
```
Comment (jmdyck) [inline spec.html:40154]:
```suggestion
        <p>The abstract operation AsyncGeneratorEnqueue takes arguments _generator_ (an AsyncGenerator), _completion_ (a Completion Record), and _promiseCapability_ (a PromiseCapability Record). It performs the following steps when called:</p>
```

File: spec.html [hunk 3]
```diff
@@ -40105,76 +40149,43 @@ <h1>AsyncGeneratorValidate ( _generator_, _generatorBrand_ )</h1>
...
-          1. Let _queue_ be _generator_.[[AsyncGeneratorQueue]].
-          1. Assert: _queue_ is not an empty List.
-          1. Let _next_ be the first element of _queue_.
-          1. Remove the first element from _queue_.
-          1. Let _promiseCapability_ be _next_.[[Capability]].
-          1. Let _iteratorResult_ be ! CreateIterResultObject(_value_, _done_).
-          1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _iteratorResult_ &raquo;).
-          1. Perform ! AsyncGeneratorResumeNext(_generator_).
-          1. Return *undefined*.
+          1. Let _realm_ be the current Realm record.
```
Comment (jmdyck) [inline spec.html:40156]:
```suggestion
          1. Let _realm_ be the current Realm Record.
```

File: spec.html [hunk 4]
```diff
@@ -40105,76 +40149,43 @@ <h1>AsyncGeneratorValidate ( _generator_, _generatorBrand_ )</h1>
...
+          1. Let _request_ be AsyncGeneratorRequest { [[Completion]]: _completion_, [[Capability]]: _promiseCapability_, [[Realm]]: _realm_ }.
+          1. Append _request_ to the end of _generator_.[[AsyncGeneratorQueue]].
         </emu-alg>
       </emu-clause>
 
-      <emu-clause id="sec-asyncgeneratorreject" aoid="AsyncGeneratorReject">
-        <h1>AsyncGeneratorReject ( _generator_, _exception_ )</h1>
-        <p>The abstract operation AsyncGeneratorReject takes arguments _generator_ and _exception_. It performs the following steps when called:</p>
+      <emu-clause id="sec-ascyngeneratorcompletestep" aoid="AsyncGeneratorCompleteStep">
```
Comment (jmdyck) [inline spec.html:40162]:
```suggestion
      <emu-clause id="sec-asyncgeneratorcompletestep" aoid="AsyncGeneratorCompleteStep">
```

File: spec.html [hunk 5]
```diff
@@ -40105,76 +40149,43 @@ <h1>AsyncGeneratorValidate ( _generator_, _generatorBrand_ )</h1>
...
+        <h1>AsyncGeneratorCompleteStep ( _generator_, _completion_, _done_ )</h1>
+        <p>The abstract operation AsyncGeneratorCompleteStep takes arguments _generator_ (an AsyncGenerator) and _completion_ (a Completion Record). It performs the following steps when called:</p>
```
Comment (jmdyck) [inline spec.html:40164]:
```suggestion
        <p>The abstract operation AsyncGeneratorCompleteStep takes arguments _generator_ (an AsyncGenerator), _completion_ (a Completion Record), and _done_ (a Boolean). It performs the following steps when called:</p>
```

File: spec.html [hunk 6]
```diff
@@ -40244,17 +40218,99 @@ <h1>AsyncGeneratorYield ( _value_ )</h1>
...
-          1. Set the code evaluation state of _genContext_ such that when evaluation is resumed with a Completion _resumptionValue_ the following steps will be performed:
-            1. If _resumptionValue_.[[Type]] is not ~return~, return Completion(_resumptionValue_).
-            1. Let _awaited_ be Await(_resumptionValue_.[[Value]]).
-            1. If _awaited_.[[Type]] is ~throw~, return Completion(_awaited_).
-            1. Assert: _awaited_.[[Type]] is ~normal~.
-            1. Return Completion { [[Type]]: ~return~, [[Value]]: _awaited_.[[Value]], [[Target]]: ~empty~ }.
-            1. NOTE: When one of the above steps returns, it returns to the evaluation of the |YieldExpression| production that originally called this abstract operation.
-          1. Return ! AsyncGeneratorResolve(_generator_, _value_, *false*).
-          1. NOTE: This returns to the evaluation of the operation that had most previously resumed evaluation of _genContext_.
+          1. Let _completion_ be NormalCompletion(value).
```
Comment (jmdyck) [inline spec.html:40221]:
```suggestion
          1. Let _completion_ be NormalCompletion(_value_).
```

File: spec.html [hunk 7]
```diff
@@ -40244,17 +40218,99 @@ <h1>AsyncGeneratorYield ( _value_ )</h1>
...
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-asyncgeneratorawaitreturn" aoid="AsyncGeneratorAwaitReturn">
+        <h1>AsyncGeneratorAwaitReturn ( _generator_ )</h1>
+        <p>The abstract operation AsyncGeneratorAwaitReturn takes argument _generator_ (an AsyncGenerator). It performs the following steps when called:</p>
+        <emu-alg>
+          1. Let _queue_ be _generator_.[[AsyncGeneratorQueue]].
+          1. Assert: _queue_ is not empty.
+          1. Let _next_ the first element of _queue_.
```
Comment (jmdyck) [inline spec.html:40246]:
```suggestion
          1. Let _next_ be the first element of _queue_.
```

File: spec.html [hunk 8]
```diff
@@ -40244,17 +40218,99 @@ <h1>AsyncGeneratorYield ( _value_ )</h1>
...
+            1. Let _completion_ be _next_.[[Completion]].
+            1. If _completion_.[[Type]] is ~return~, then
+              1. Set _generator_.[[AsyncGeneratorState]] to ~awaiting-return~.
+              1. Perform ! AsyncGeneratorAwaitReturn(_generator_).
+              1. Set _done_ to *true*.
+            1. Else,
+              1. If _completion_.[[Type]] is ~normal~, then
+                1. Set _completion_ to NormalCompletion(*undefined*).
+              1. Perform ! AsyncGeneratorCompleteStep(_generator_, _completion_, *true*).
+              1. If _queue_ is empty, set *done* to *true*.
```
Comment (jmdyck) [inline spec.html:40313]:
```suggestion
              1. If _queue_ is empty, set _done_ to *true*.
```

**Review by jmdyck — CHANGES_REQUESTED** [review]

**Comment by jmdyck** [comment]
Also, this PR undefines the following ids:
- sec-asyncgeneratorreject
- sec-asyncgeneratorresolve
- sec-asyncgeneratorresumenext

(I don't know if there's an appropriate place to make them oldids.)

File: spec.html (hunk 1 above)
Comment (bakkot) [reply spec.html:39984]: Fixed.

**Comment by bakkot** [comment]
> Also, this PR undefines the following ids:

Yeah, I don't think there's a reasonable place to put them. They don't really correspond to any of the new operations - the whole point of the PR is that those previous operations were not a clean way of organizing the logic here, after all.

File: spec.html [hunk 9]
```diff
@@ -40244,17 +40215,102 @@ <h1>AsyncGeneratorYield ( _value_ )</h1>
...
-            1. Let _awaited_ be Await(_resumptionValue_.[[Value]]).
-            1. If _awaited_.[[Type]] is ~throw~, return Completion(_awaited_).
-            1. Assert: _awaited_.[[Type]] is ~normal~.
-            1. Return Completion { [[Type]]: ~return~, [[Value]]: _awaited_.[[Value]], [[Target]]: ~empty~ }.
-            1. NOTE: When one of the above steps returns, it returns to the evaluation of the |YieldExpression| production that originally called this abstract operation.
-          1. Return ! AsyncGeneratorResolve(_generator_, _value_, *false*).
-          1. NOTE: This returns to the evaluation of the operation that had most previously resumed evaluation of _genContext_.
+          1. Let _completion_ be NormalCompletion(_value_).
+          1. Assert: There are at least two items on the execution context stack.
+          1. Let _previousContext_ be the execution context immediately beneath the running execution context.
```
Comment (jmdyck) [inline spec.html:40220]:
```suggestion
          1. Assert: The execution context stack has at least two elements.
          1. Let _previousContext_ be the second to top element of the execution context stack.
```
(precedent in [19.2.1 eval(x)](https://tc39.es/ecma262/#sec-eval-x) and [20.2.1.1.1 CreateDynamicFunction](https://tc39.es/ecma262/#sec-createdynamicfunction))

**Review by syg — APPROVED** [review]
@michaelficarra and I are splitting up reviewing the current text for correspondence with [the gist](https://gist.github.com/bakkot/82186e27a2e78c6e507c56dd49d8a197) and reviewing the PR text for correspondence.

This lgtm is for correspondence with the current text.

The state machine in the gist reads correct to me. My understanding is as follows.

The current spec uses AsyncGeneratorResumeNext as a message pump that ~~blocks~~ suspends to wait for new messages. That is, it doesn't just resume next and continues execution, it generally pumps until it runs out of messages. This pump loop is awkwardly captured by self-recursive calls in the following states:

- From `executing`, when a `yield` is evaluated, if the queue is not empty, there is a synchronous transition `executing` -> `suspendedYield` -> `executing`.
- From `completed`, the queue is pumped and each promise is resolved with undefined until a return completion, in which case it goes into `awaiting-return`.

Put a third way, the current spec puts the focus on the message pump machinery, which is more confusing than putting the focus on the state machine machinery.

File: spec.html [hunk 10]
```diff
@@ -40244,17 +40215,102 @@ <h1>AsyncGeneratorYield ( _value_ )</h1>
...
+          1. Set _onFulfilled_.[[Generator]] to _generator_.
+          1. Let _stepsRejected_ be the algorithm steps defined in <emu-xref href="#async-generator-await-return-rejected-functions" title></emu-xref>.
+          1. Let _lengthRejected_ be the number of non-optional parameters of the function definition in <emu-xref href="#async-generator-await-return-rejected-functions" title></emu-xref>.
+          1. Let _onRejected_ be ! CreateBuiltinFunction(_stepsRejected_, _lengthRejected_, *""*, &laquo; [[Generator]] &raquo;).
+          1. Set _onRejected_.[[Generator]] to _generator_.
+          1. Perform ! PerformPromiseThen(_promise_, _onFulfilled_, _onRejected_).
+        </emu-alg>
+
+        <emu-clause id="async-generator-await-return-fulfilled-functions" oldids="async-generator-resume-next-return-processor-fulfilled">
+          <h1>AsyncGeneratorAwaitReturn Fulfilled Functions</h1>
```
Comment (michaelficarra) [inline spec.html:40262]: I'd prefer to finish up #2109 and then make these abstract closures.

File: spec.html [hunk 11]
```diff
@@ -40244,17 +40215,102 @@ <h1>AsyncGeneratorYield ( _value_ )</h1>
...
+          <p>An AsyncGeneratorAwaitReturn fulfilled function is an anonymous built-in function that is used as part of the AsyncGeneratorAwaitReturn specification device to unwrap promises passed in to the <emu-xref href="#sec-asyncgenerator-prototype-return" title></emu-xref> method. Each AsyncGeneratorAwaitReturn fulfilled function function has a [[Generator]] internal slot.</p>
```
Comment (michaelficarra) [inline spec.html:40263]:
```suggestion
          <p>An AsyncGeneratorAwaitReturn fulfilled function is an anonymous built-in function that is used as part of the AsyncGeneratorAwaitReturn specification device to unwrap promises passed in to the <emu-xref href="#sec-asyncgenerator-prototype-return" title></emu-xref> method. Each AsyncGeneratorAwaitReturn fulfilled function has a [[Generator]] internal slot.</p>
```

**Review by michaelficarra — APPROVED** [review]
LGTM with respect to correspondence to the state machine.

**Pull request merged**